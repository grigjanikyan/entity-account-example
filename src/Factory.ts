import { AccountId, AccountRecord } from '@S_E_C_R_E_T/types';
import Modularity, { TAnyValue } from '@S_E_C_R_E_T/modularity';

import { AccountDal } from './Dal';
import { AccountService } from './Service';
import { AccountTransform } from './Transform';
import { AccountController } from './Controller';
import { AccountValidator } from './Validator';
import { AccountValidation } from './Validation';

export * from './Dal';
export { AccountController } from './Controller';

export function accountFactory(
  params: TAnyValue,
): Modularity<
  AccountDal<AccountId, AccountRecord>,
  AccountService<AccountId, AccountRecord>,
  AccountTransform<AccountId, AccountRecord>,
  AccountController<AccountId, AccountRecord>,
  AccountValidator<AccountId, AccountRecord>,
  AccountValidation<AccountId, AccountRecord>
> {
  return new Modularity<
    AccountDal<AccountId, AccountRecord>,
    AccountService<AccountId, AccountRecord>,
    AccountTransform<AccountId, AccountRecord>,
    AccountController<AccountId, AccountRecord>,
    AccountValidator<AccountId, AccountRecord>,
    AccountValidation<AccountId, AccountRecord>
  >({
    params: {
      tableName: 'account',
      fields: [
        'id',
        'version',
        'createdAt',
        'createdBy',
        'updatedAt',
        'updatedBy',
        'removedAt',
        'removedBy',
        'email',
        'password',
        'active',
        'firstName',
        'lastName',
        'mobilePhone',
        'roles',
      ],
      ...params,
    },
    userLayers: {
      Dal: AccountDal,
      Service: AccountService,
      Transform: AccountTransform,
      Controller: AccountController,
      Validator: AccountValidator,
      Validation: AccountValidation,
    },
  });
}
