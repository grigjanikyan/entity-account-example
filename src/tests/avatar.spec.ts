import { SessionJWT, SessionJWTPayload } from '@S_E_C_R_E_T/types';
import { IUserSessionRequest, IPromiseResponse } from '@S_E_C_R_E_T/middlewares';

import { PG as PgStub } from '@S_E_C_R_E_T/db-pg';

import { accountFactory as factory } from '../Factory';

describe('Account avatar', () => {
  it('upload avatar', async () => {
    const queryRunner = new PgStub([[{ id: 222, avatar: 'avatar-uuid-old' }], [{ id: 222 }]]);
    const generateUuid = jest.fn().mockReturnValue('avatar-uuid');
    const avatarStorage = {
      uploadBuffer: jest.fn().mockResolvedValue({ Location: 'url-to-avatar' }),
      remove: jest.fn().mockResolvedValue({}),
    };
    let promise = Promise.resolve();
    // eslint-disable-next-line no-return-assign
    const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
    const accountFactory = factory({
      queryRunner,
      generateUuid,
      avatarStorage,
      getCurrentDateTime: () => 'today or yesterday who knows',
    });
    const buffer = Buffer.alloc(1000, 0);
    Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]).copy(buffer);
    accountFactory.controller.uploadAvatar(
      ({
        params: { id: '111' },
        file: { buffer },
        getUser: () => ({ id: 222, roles: [] }),
      } as unknown) as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
    );
    const serviceResult = await promise;
    expect(queryRunner.queries).toMatchSnapshot();
    expect(avatarStorage.remove).toBeCalledWith('avatar-uuid-old');
    expect(avatarStorage.uploadBuffer).toBeCalledWith({
      fileName: 'avatar-uuid',
      buffer,
      mimeType: 'image/png',
      publicAccess: true,
      tags: [{ Key: 'accountId', Value: '222' }],
    });
    expect(serviceResult).toStrictEqual({ id: 222, avatar: 'url-to-avatar' });
  });

  it('validation check', async () => {
    const queryRunner = new PgStub([[{ id: 222, avatar: 'avatar-uuid-old' }], [{ id: 222 }]]);
    const generateUuid = jest.fn().mockReturnValue('avatar-uuid');
    const avatarStorage = {
      uploadBuffer: jest.fn().mockResolvedValue({ Location: 'url-to-avatar' }),
      remove: jest.fn().mockResolvedValue({}),
    };
    let promise = Promise.resolve();
    // eslint-disable-next-line no-return-assign
    const res = ({ promise: jest.fn((prom) => (promise = prom)) } as unknown) as IPromiseResponse;
    const accountFactory = factory({
      queryRunner,
      generateUuid,
      avatarStorage,
      getCurrentDateTime: () => 'today or yesterday who knows',
    });
    const buffer = Buffer.alloc(1000, 0);
    Buffer.from([0x89, 0x5, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]).copy(buffer);
    accountFactory.controller.uploadAvatar(
      ({
        params: { id: '111' },
        file: { buffer },
        getUser: () => ({ id: 222, roles: [] }),
      } as unknown) as IUserSessionRequest<SessionJWTPayload, SessionJWT>,
      res,
    );
    try {
      await promise;
    } catch (error) {
      return expect(error.fields).toStrictEqual({ image: 'Неправильный файл изображения!' });
    }
    throw new Error('Should not be reached');
  });
});
