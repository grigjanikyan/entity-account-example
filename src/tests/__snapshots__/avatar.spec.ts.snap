// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account avatar upload avatar 1`] = `
Array [
  Object {
    "bindings": Array [
      222,
    ],
    "sql": "select \\"id\\", \\"version\\", \\"created_at\\", \\"created_by\\", \\"updated_at\\", \\"updated_by\\", \\"removed_at\\", \\"removed_by\\", \\"email\\", \\"password\\", \\"active\\", \\"first_name\\", \\"last_name\\", \\"mobile_phone\\", \\"roles\\" from \\"account\\" where \\"id\\" = $1",
  },
  Object {
    "bindings": Array [
      1,
      "today or yesterday who knows",
      222,
      222,
    ],
    "sql": "update \\"account\\" set \\"version\\" = $1, \\"updated_at\\" = $2, \\"updated_by\\" = $3 where \\"id\\" = $4 returning \\"id\\"",
  },
]
`;
